program/src/processor/sol_staker_stake_tokens.rs

```diff
@@ -0,0 +1,184 @@
+
+    require!(
+        ctx.accounts.sol_staker_stake.owner == program_id,
+        ProgramError::InvalidAccountOwner,
+        "validator stake"
+    );
+
+    let mut sol_staker_stake_data = ctx.accounts.sol_staker_stake.try_borrow_mut_data()?;
+    let sol_staker_stake = unpack_initialized_mut::<SolStakerStake>(&mut sol_staker_stake_data)?;
+
+    let (derivation, _) = find_sol_staker_stake_pda(
+        &sol_staker_stake.sol_stake,
+        ctx.accounts.config.key,
+        program_id,
+    );
+
+    require!(
+        ctx.accounts.sol_staker_stake.key == &derivation,
+        ProgramError::InvalidSeeds,
+        "sol staker stake",
+    );
+
+    // validator stake
+    // - owner must be the stake program
+    // - must be initialized
+    // - must have the correct derivation (validator vote must match the validator vote in
+    //   the sol staker stake account)
+
+    require!(
+        ctx.accounts.validator_stake.owner == program_id,
+        ProgramError::InvalidAccountOwner,
+        "validator stake"
+    );
+
+    let mut validator_stake_data = ctx.accounts.validator_stake.try_borrow_mut_data()?;
+    let validator_stake = unpack_initialized_mut::<ValidatorStake>(&mut validator_stake_data)?;
+
+    let (derivation, _) = find_validator_stake_pda(
+        &sol_staker_stake.delegation.validator_vote,
+        ctx.accounts.config.key,
+        program_id,
+    );
+
+    require!(
+        ctx.accounts.validator_stake.key == &derivation,
+        ProgramError::InvalidSeeds,
+        "validator stake",
+    );
+
+    // vault
+    // - must be the token account on the stake config account
+
+    require!(
+        ctx.accounts.vault.key == &config.vault,
+        StakeError::IncorrectVaultAccount,
+    );
+
+    let vault_data = ctx.accounts.vault.try_borrow_data()?;
+    // unpack to validate the mint
+    let vault = PodStateWithExtensions::<PodAccount>::unpack(&vault_data)?;
+
+    // mint
+    // - must match the stake vault mint
+
+    require!(
+        &vault.base.mint == ctx.accounts.mint.key,
+        StakeError::InvalidMint,
+        "mint"
+    );
+
+    let mint_data = ctx.accounts.mint.try_borrow_data()?;
+    let mint = PodStateWithExtensions::<PodMint>::unpack(&mint_data)?;
+    let decimals = mint.base.decimals;
+
+    // Update the config and stake account data.
+
+    require!(amount > 0, StakeError::InvalidAmount);
+
+    let updated_staked_amount = sol_staker_stake
+        .delegation
+        .amount
+        .checked_add(amount)
+        .ok_or(ProgramError::ArithmeticOverflow)?;
+    // maximum allowed stake based on the SOL amount
+    let limit = calculate_maximum_stake_for_sol_amount(sol_staker_stake.lamports_amount)?;
+
+    require!(
+        updated_staked_amount as u128 <= limit,
+        StakeError::TotalStakeAmountExceedsSolLimit,
+        "current staked amount ({}) + new amount ({}) exceeds limit ({})",
+        sol_staker_stake.delegation.amount,
+        amount,
+        limit
+    );
```

Comment:
So just to make sure, the idea is that syncing between the SOL stake and PAL stake only ever happens in `Sync`, correct? If so, sounds great!

Reply:
Agreed from my side, let's keep it simple and only do it in `Sync`.

Reply:
Cool!